memtable/write_buffer_manager.cc

```diff
@@ -17,37 +17,7 @@ namespace ROCKSDB_NAMESPACE {
 #ifndef ROCKSDB_LITE
 namespace {
 const size_t kSizeDummyEntry = 256 * 1024;
-// The key will be longer than keys for blocks in SST files so they won't
-// conflict.
-const size_t kCacheKeyPrefix = kMaxVarint64Length * 4 + 1;
 }  // namespace
-
-struct WriteBufferManager::CacheRep {
```

Comment:
[non-blocking]
Question 3: just a general question on WBM to understand the code better: why do we need things like "std::lock_guard<std::mutex> lock(cache_rep_->cache_mutex_);" at line 100? What concurrency issue we are trying to prevent here?

Reply:
> above in here #2350 (comment). But I still can't get any answer from the discussion there.

That tends to be the case with discussions involving me. Trying again. I think functions like ReserveMemWithCache cannot run in parallel because of updates to unprotected structure like vector. There is also some complexity with making sure updates across variables are atomic. It could probably be done with more fine-grained locking if needed.

The linked comment I suggested an (important) edge case might work already with the locking removed, not the general case.

Unrelated: does FreeMemWithCache release at most one dummy entry? I feel there should be a loop.

Reply:
> Unrelated: does FreeMemWithCache release at most one dummy entry? I feel there should be a loop.

Maybe that's why we have to free dummy entries in the destructor, because we're skipping most of them when they're actually freed.

Also this supports your design (2) idea so we do not have to duplicate these logic as it's evidently error-prone.